src/IO/S3/Client.cpp

```diff
@@ -421,9 +456,23 @@ Client::doRequest(const RequestType & request, RequestFn request_fn) const
     throw Exception(ErrorCodes::TOO_MANY_REDIRECTS, "Too many redirects");
 }
 
-ProviderType Client::getProviderType() const
+bool Client::supportsMultiPartCopy() const
 {
-    return provider_type;
+    return provider_type != ProviderType::GCS;
+}
```

Comment:
Nitpicking: Unrelated to this PR but worth understanding some implications when using GCS. Why do we disable multipart upload in GCS? I thought this was already supported and the "only" missing big difference between GCS and S3 was the multi-delete.

> This page discusses XML API multipart uploads in Cloud Storage. This upload method uploads files in parts and then assembles them into a single object using a final request. XML API multipart uploads are compatible with Amazon S3 multipart uploads.

https://cloud.google.com/storage/docs/multipart-uploads

Reply:
We don't disable multipart upload but multipart **native copy**.
This is not supported by GCS so we have to try to do single part **native copy** and if it fails fallback to upload instead of multipart native copy as we do for AWS.
You can read some things left to do in the previous PR description https://github.com/ClickHouse/ClickHouse/pull/48981

The only difference is that GCS support team confirmed that they don't have a limit of 5GB for a single part native copy so we will probably remove the limit if GCS is used.
Additional problem is the need to [compose](https://cloud.google.com/storage/docs/xml-api/put-object-compose) a file uploaded with multi part upload before it can be copied natively but it returns 503 in some cases. This was also confirmed to be a problem on their side but I prepared a PR for it and will push it once it's resolved.

Reply:
Until all the improvements are made it's important to be aware that backup **could** be slower than AWS in some cases.

Reply:
Great, thank you @antonio2368!

Reply:
@antonio2368 maybe worth to auto-detect this feature, like `DeleteObjects`?
https://github.com/ClickHouse/ClickHouse/blob/353791b2dd9e7f309f2a4b300c701cd32c1bf9df/src/Disks/ObjectStorages/S3/registerDiskS3.cpp#L41

Since there can be other provides that will not support some features.

Reply:
yes, that sounds like a better solution
maybe combine it and do it if the provider type is UNKNOWN because we already know best on URL for AWS and GCS